File: 0567-permutation-in-string/0567-permutation-in-string.py
from collections import Counter
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1)>len(s2):
            return False
        need=Counter()
        count=Counter()
        l=0
        for i in s1:
            need[i]=1+need.get(i,0)
        for r in range(len(s1)):
            count[s2[r]]=1+count.get(s2[r],0)
        for r in range(len(s1),len(s2)):
            if count==need:
                return True
            
            count[s2[r]]=1+count.get(s2[r],0)
            count[s2[l]]-=1
            l+=1
        if count==need:
            return True
        return False

File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkIt(int a[],int b[]){
        for(int i=0;i<26;i++){
            if(a[i]!=b[i]) return false;
        }
        return true;
    }

    bool checkInclusion(string s1, string s2) {
        if(s1.length()>s2.length()) return false;
        int need[26]={0};
        int index=0;
        for(int i=0;i<s1.length();i++){
            index=s1[i]-'a';
            need[index]++;
        }
        int l=0;
        int count[26]={0};
        for(int i=0;i<s1.length();i++){
            index=s2[i]-'a';
            count[index]++;
        }
        for(int r=s1.length();r<s2.length();r++){
            if(checkIt(need,count)) return true;
            index=s2[r]-'a';
            count[index]++;
            index=s2[l]-'a';
            count[index]--;
            l++;
        }
        if(checkIt(need,count)) return true;
        return false;
    }
};